source/agora/utils/HumanReadable.d

```diff
@@ -0,0 +1,284 @@
+    It is assumed that a human-readable representation will not rely on
+    non-visible data, or data that cannot be accessed without mutation.
+
+    Copyright:
+        Copyright (c) 2019 BOS Platform Foundation Korea
+        All rights reserved.
+
+    License:
+        MIT License. See LICENSE for details.
+
+*******************************************************************************/
+
+module agora.utils.HumanReadable;
+
+import agora.common.Amount;
+import agora.common.Data;
+import agora.common.crypto.Key;
+import agora.consensus.data.Block;
+import agora.consensus.data.Transaction;
+
+import std.algorithm;
+import std.format;
+import std.range;
+
+/// Ditto
+public auto humanReadable (T) (const ref T input)
+{
+    static if (is(T : const Amount))
+        return AmountFmt(input);
+    else static if (is(T : const Amount))
+        return HashFmt(input);
+    else static if (is(T : const Input))
+        return InputFmt(input);
+    else static if (is(T : const Output))
+        return OutputFmt(input);
+    else static if (is(T : const Transaction))
+        return TransactionFmt(input);
+    else static if (is(T : const Block))
+        return BlockFmt(input);
+    else
+        return input;
+}
+
+/// Formatting struct for `Amount`
+private struct AmountFmt
+{
+    private Amount value;
+
+    public void toString (scope void delegate(scope const(char)[]) @safe sink) @safe
+    {
+        formattedWrite(sink, "%,d.%,.8d", this.value.integral(), this.value.decimal());
+    }
+}
+
+@safe unittest
+{
+    immutable one = Amount(1);
+    assert(format("%s", AmountFmt(Amount.MaxUnitSupply)) == "500,000,000.0000,000");
+    assert(format("%s", AmountFmt(one)) == "0.0000,001");
+    assert(format("%s", AmountFmt(Amount.UnitPerCoin)) == "1.0000,000");
+}
+
+/// Formatting struct for `Hash` and `Signature`
+private struct HashFmt
+{
+    private const(Hash) value;
+
+    public this (ref const Hash r) @safe
+    {
+        this.value = r;
+    }
+
+    public void toString (scope void delegate(scope const(char)[]) @safe sink) @safe
+    {
+        // Only format `0xABCD..EFGH`
+        enum StartUntil = 6;
+        enum EndFrom    = Hash.StringBufferSize - 4;
+        size_t count;
+        scope void delegate(scope const(char)[]) @safe wrapper = (scope data) @safe {
+                if (count < StartUntil)
+                {
+                    sink(data);
+                    if (count + data.length >= StartUntil)
+                        sink("...");
+                }
+                if (count >= EndFrom)
+                    sink(data);
+                count += data.length;
+            };
+        this.value.toString(wrapper);
+    }
+}
+
+@safe unittest
+{
+    static immutable Hash SomeHash =
+        "0x000000000019d6689c085ae165831e934ff763ae46a2a6c172b3f1b60a8ce26f"
+        ~ "000000000019d6689c085ae165831e934ff763ae46a2a6c172b3f1b60a8ce26f";
+    assert(format("%s", HashFmt(SomeHash)) == "0x0000...e26f");
```

Comment:
Do you think we could make it all-caps? I'm thinking it might improve the readability.

Reply:
Ah, actually you are using all-caps for public keys. It makes sense to keep this lowercase.